File: 56-merge-intervals/merge-intervals.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.LinkedList;

class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0]));
        LinkedList<int[]> links = new LinkedList<>();

        for (int[] interval: intervals) {
            int x1 = interval[0];
            int y2 = interval[1];
            if (links.isEmpty()) {
                links.add(interval);
            } else {
                int[] prev = links.getLast();
                int x = prev[0];
                int y = prev[1];
                if (x1 > y) {
                    links.add(interval);
                } else {
                    links.getLast()[1] = Math.max(y2, y);
                }
            }
        }
        return links.toArray(new int[links.size()][]);
    }
}

File: 56-merge-intervals/merge-intervals.swift
class Solution {
    func merge(_ intervals: [[Int]]) -> [[Int]] {
        var ints: [[Int]] = intervals.sorted{$0[0] < $1[0]}, result: [[Int]] = []
        var x: Int = ints[0][0], y = ints[0][1]

        for i: Int in 1..<ints.count {
            if ints[i][0] > y {
                result.append([x, y])
                x = ints[i][0]; y = ints[i][1]
            } else {
                y = max(y, ints[i][1])
            }
        }
        result.append([x, y])
        return result
    }
}